internal/ingest/download_sip.go

```diff
@@ -0,0 +1,97 @@
+package ingest
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+
+	"github.com/google/uuid"
+	"gocloud.dev/gcerrors"
+
+	goaingest "github.com/artefactual-sdps/enduro/internal/api/gen/ingest"
+	"github.com/artefactual-sdps/enduro/internal/datatypes"
+	"github.com/artefactual-sdps/enduro/internal/persistence"
+)
+
+func (w *goaWrapper) checkSIP(ctx context.Context, id string) (*datatypes.SIP, error) {
+	// Validate the payload UUID.
+	sipUUID, err := uuid.Parse(id)
+	if err != nil {
+		return nil, goaingest.MakeNotValid(errors.New("invalid UUID"))
+	}
+
+	// Read the persisted SIP.
+	sip, err := w.perSvc.ReadSIP(ctx, sipUUID)
+	if err != nil {
+		if errors.Is(err, persistence.ErrNotFound) {
+			return nil, &goaingest.SIPNotFound{UUID: id, Message: "SIP not found"}
+		} else {
+			return nil, goaingest.MakeInternalError(errors.New("error reading SIP"))
+		}
+	}
+
+	// Check that failed as and failed key values are set.
+	if sip.FailedAs == "" || sip.FailedKey == "" {
+		return nil, goaingest.MakeNotValid(errors.New("SIP has no failed values"))
+	}
+
+	return sip, nil
+}
+
+func (w *goaWrapper) DownloadSipRequest(
+	ctx context.Context,
+	payload *goaingest.DownloadSipRequestPayload,
+) (*goaingest.DownloadSipRequestResult, error) {
+	// Check the SIP status.
+	_, err := w.checkSIP(ctx, payload.UUID)
+	if err != nil {
+		return nil, err
+	}
+
+	// Request a ticket.
+	ticket, err := w.ticketProvider.Request(ctx)
+	if err != nil {
+		return nil, goaingest.MakeInternalError(errors.New("failed to request ticket"))
```

Comment:
I think "ticket request failed" or "failed to receive ticket" would be a more accurate error message.